hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+  }
+
+  for (int i = 0; i < in.size(); ++i) {
+    state[i] = state[i] ^ key[i];
+  }
+
+  return state;
+}
+
+// Apply a full set of substitution/permutation rounds for decrypt to the
+// incoming byte vector
+static std::vector<uint8_t> scramble_subst_perm_dec(
+    const std::vector<uint8_t> &in, const std::vector<uint8_t> &key,
+    uint32_t bit_width, uint32_t num_rounds) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  assert(key.size() >= ((bit_width + 7) / 8));
+
+  std::vector<uint8_t> state(in);
+
+  for (int i = 0; i < num_rounds; ++i) {
+    for (int j = 0; j < in.size(); ++j) {
+      state[j] = state[j] ^ key[j];
+    }
+
+    state = scramble_perm_layer(state, bit_width, true);
+    state = scramble_flip_layer(state, bit_width);
+    state = scramble_sbox_layer(state, bit_width, true);
+  }
+
+  for (int i = 0; i < in.size(); ++i) {
+    state[i] = state[i] ^ key[i];
+  }
+
+  return state;
+}
+
+// Generate a keystream for XORing with data using PRINCE. Multiple PRINCE
+// instantiatons are used when the keystream is greater than a single PRINCE
+// width (64-bit).
+static std::vector<uint8_t> scramble_gen_keystream(
+    const std::vector<uint8_t> &addr, uint32_t addr_width,
+    const std::vector<uint8_t> &nonce, const std::vector<uint8_t> &key,
+    uint32_t keystream_width, uint32_t num_half_rounds) {
+  assert(key.size() == (kPrinceWidthByte * 2));
+
+  // Determine how many PRINCE replications are required
+  uint32_t num_prince = (keystream_width + kPrinceWidth - 1) / kPrinceWidth;
+
+  std::vector<uint8_t> keystream;
+
+  for (int i = 0; i < num_prince; ++i) {
+    // Initial vector is data for PRINCE to encrypt. Formed from nonce and data
+    // address
+    std::vector<uint8_t> iv(8, 0);
+
+    for (int j = 0; j < kPrinceWidth; ++j) {
+      if (j < addr_width) {
+        // Bottom addr_width bits of IV are address
+        or_vector_bit(iv, j, read_vector_bit(addr, j));
+      } else {
+        // Other bits are taken from nonce. Each PRINCE instaniations will used
+        // different nonce bits.
+        or_vector_bit(
+            iv, j,
+            read_vector_bit(
+                nonce, (j - addr_width) + i * (kPrinceWidth - addr_width)));
+      }
+    }
+
+    // PRINCE C reference model works on big-endian byte order
+    iv = byte_swap_vector(iv);
+    std::vector<uint8_t> key_be = byte_swap_vector(key);
+
+    // Apply PRINCE to IV to produce keystream
+    std::vector<uint8_t> keystream_block(kPrinceWidthByte);
+    prince_enc_dec(&iv[0], &key_be[0], &keystream_block[0], 0, num_half_rounds,
+                   0);
+
+    // Flip keystream into little endian order and add to keystream vector
+    keystream_block = byte_swap_vector(keystream_block);
+    keystream.insert(keystream.end(), keystream_block.begin(),
+                     keystream_block.end());
+  }
+
+  return keystream;
+}
+
+// Split incoming data into `subst_perm_width` chunks and individually apply the
```

Comment:
Nit: The other comments in the file aren't using backticks to quote symbol names. I'm not sure there's any need to here either: maybe drop them for consistency?